catalog/core/catalog-core-api-impl/src/test/groovy/ddf/catalog/util/impl/ResultIterableSpec.groovy

```diff
@@ -0,0 +1,372 @@
+import ddf.catalog.operation.Query
+import ddf.catalog.operation.QueryRequest
+import ddf.catalog.operation.QueryResponse
+import ddf.catalog.source.SourceUnavailableException
+import ddf.catalog.source.UnsupportedQueryException
+import spock.lang.Specification
+import spock.lang.Unroll
+
+class QueryResultPaginatorSpec extends Specification {
+
+    CatalogFramework catalogFramework
+
+    def setup() {
+        catalogFramework = Mock(CatalogFramework.class)
+    }
+
+    def "hasNext() is false when catalog returns no results"() {
+        setup:
+        1 * catalogFramework.query(_ as QueryRequest) >> {
+            QueryRequest queryRequest -> buildQueryResponse(queryRequest, 0, 0)
+        }
+
+        Query queryMock = createQueryMock(1, 1)
+        QueryRequest queryRequestMock = createQueryRequestMock(queryMock)
+
+        def resultIterator = new ResultIterable(catalogFramework, queryRequestMock).iterator()
+
+        when:
+        def hasNext = resultIterator.hasNext()
+
+        then:
+        hasNext.is false
+    }
+
+    def "hasNext() is true when catalog returns results"() {
+        setup:
+        1 * catalogFramework.query(_ as QueryRequest) >> {
+            QueryRequest queryRequest -> buildQueryResponse(queryRequest, 1, 1)
+        }
+
+        Query queryMock = createQueryMock(1, 1)
+        QueryRequest queryRequestMock = createQueryRequestMock(queryMock)
+
+        def resultIterator = new ResultIterable(catalogFramework, queryRequestMock).iterator()
+
+        when:
+        def hasNext = resultIterator.hasNext()
+
+        then:
+        hasNext.is true
+    }
+
+    def "hasNext() is true when called twice in a row and catalog returns results"() {
+        setup:
+        1 * catalogFramework.query(_ as QueryRequest) >> {
+            QueryRequest queryRequest -> buildQueryResponse(queryRequest, 1, 1)
+        }
+
+        Query queryMock = createQueryMock(1, 1)
+        QueryRequest queryRequestMock = createQueryRequestMock(queryMock)
+
+        def resultIterator = new ResultIterable(catalogFramework, queryRequestMock).iterator()
+
+        when:
+        resultIterator.hasNext()
+        def hasNext = resultIterator.hasNext()
+
+        then:
+        hasNext.is true
+    }
+
+    def "hasNext() is true when using a query function and catalog returns results"() {
+        setup:
+        1 * catalogFramework.query(_ as QueryRequest) >> {
+            QueryRequest queryRequest -> buildQueryResponse(queryRequest, 1, 1)
+        }
+
+        Query queryMock = createQueryMock(1, 1)
+        QueryRequest queryRequestMock = createQueryRequestMock(queryMock)
+
+        def resultIterator = new ResultIterable(catalogFramework.&query, queryRequestMock).iterator()
+
+        when:
+        def hasNext = resultIterator.hasNext()
+
+        then:
+        hasNext.is true
+    }
+
+    @Unroll
+    def '''test that #nextCalls results are returned and catalog is queried #expectedQueries times
```

Comment:
Ryan loved the data table (I do too), but I'm going to be honest and say I might love this long, fully descriptive test name even more. 👍